pkg/controller/sync/sync_controller.go

```diff
@@ -39,56 +37,53 @@ import (
 var log = logf.Log.WithName("controller").WithValues("metaKind", "Sync")
 
 type Adder struct {
-	Opa *opa.Client
+	Opa    OpaDataClient
+	Events <-chan event.GenericEvent
 }
 
 // Add creates a new Sync Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
 // and Start it when the Manager is Started.
-func (a *Adder) Add(mgr manager.Manager, gvk schema.GroupVersionKind, cs *watch.ControllerSwitch) error {
-	r := newReconciler(mgr, gvk, a.Opa, cs)
-	return add(mgr, r, gvk)
+func (a *Adder) Add(mgr manager.Manager) error {
+	r := newReconciler(mgr, a.Opa)
+	return add(mgr, r, a.Events)
 }
 
 // newReconciler returns a new reconcile.Reconciler
-func newReconciler(mgr manager.Manager, gvk schema.GroupVersionKind, opa *opa.Client, cs *watch.ControllerSwitch) reconcile.Reconciler {
+func newReconciler(mgr manager.Manager, opa OpaDataClient) reconcile.Reconciler {
 	return &ReconcileSync{
-		Client: mgr.GetClient(),
-		cs:     cs,
+		reader: mgr.GetCache(),
 		scheme: mgr.GetScheme(),
 		opa:    opa,
-		log:    log.WithValues("kind", gvk.Kind, "apiVersion", gvk.GroupVersion().String()),
-		gvk:    gvk,
+		log:    log,
 	}
 }
 
 // add adds a new Controller to mgr with r as the reconcile.Reconciler
-func add(mgr manager.Manager, r reconcile.Reconciler, gvk schema.GroupVersionKind) error {
+func add(mgr manager.Manager, r reconcile.Reconciler, events <-chan event.GenericEvent) error {
 	// Create a new controller
-	c, err := controller.New(fmt.Sprintf("%s-sync-controller", gvk.String()), mgr, controller.Options{Reconciler: r})
+	c, err := controller.New("sync-controller", mgr, controller.Options{Reconciler: r})
 	if err != nil {
 		return err
 	}
 
 	// Watch for changes to the provided resource
-	instance := unstructured.Unstructured{}
-	instance.SetGroupVersionKind(gvk)
-	err = c.Watch(&source.Kind{Type: &instance}, &handler.EnqueueRequestForObject{})
-	if err != nil {
-		return err
-	}
-
-	return nil
+	return c.Watch(
+		&source.Channel{
+			Source:         events,
+			DestBufferSize: 1024,
+		},
+		&handler.EnqueueRequestsFromMapFunc{ToRequests: util.EventPacker{}},
+	)
 }
 
 var _ reconcile.Reconciler = &ReconcileSync{}
 
 // ReconcileSync reconciles an arbitrary object described by Kind
 type ReconcileSync struct {
-	client.Client
-	cs     *watch.ControllerSwitch
+	reader client.Reader
+
```

Comment:
whats the reason for removing this for sync?

Reply:
The ControllerSwitch was only needed in controllers that would fight conflict with the teardown logic in `main.go:L194`. The sync controller isn't writing status or removing finalizers so there shouldn't be any contention to worry about.